File: python_practice/1300.py
import sys
input = sys.stdin.readline

n, k = [int(input()) for _ in range(2)]
l, r = 1, n**2

ans = -1

while l <= r:
    mid = l + (r-l) // 2
    cnt = 0

    for el in range(1, n+1):
        cnt += min(mid // el, n)
    
    if cnt >= k:
        ans = mid
        r = mid - 1
    else:
        l = mid + 1

print(ans)

File: BOJ/1300.cpp
#include<iostream>

using namespace std;

typedef unsigned long long ull;

int main(void) {
	ull n, k;
	cin >> n >> k;
	ull low=1, high=k;
	ull ans;
	while(low<=high) {
		ull mid = low+(high-low)/2;
		cout << low << " " << high << " " << mid << '\n';
		ull cnt=0;
		for(int i=1; i<=n; ++i) cnt+=min(mid/i, n);
		if(cnt<=k){
			ans=mid;
			low=mid+1;
		} else high=mid-1;
	}
	cout << ans << '\n';
}